src/main/2.4/scala/com/crealytics/spark/v2/excel/ExcelDataSource.scala

```diff
@@ -256,36 +256,48 @@ class ExcelDataSourceReader(
       sample = if (sample < 1) 1 else sample
       inputPaths.take(sample).map(_.getPath.toUri)
     }
-    var rows = excelHelper.getRows(conf, paths.head)
+    var rows = excelHelper.getSheetData(conf, paths.head)
 
-    if (rows.isEmpty) { /* If the first file is empty, not checking further */
+    if (rows.rowIterator.isEmpty) { /* If the first file is empty, not checking further */
       StructType(Seq.empty)
     } else {
-      /* Prepare field names */
-      val colNames =
-        if (options.header) { /* Get column name from the first row */
-          val r = excelHelper.getColumnNames(rows.next)
-          rows = rows.drop(options.ignoreAfterHeader)
-          r
-        } else { /* Peek first row, then return back */
-          val headerRow = rows.next
-          val r = excelHelper.getColumnNames(headerRow)
-          rows = Iterator(headerRow) ++ rows
-          r
+      try {
+        /* Prepare field names */
+        val colNames =
+          if (options.header) {
+            /* Get column name from the first row */
+            val r = excelHelper.getColumnNames(rows.rowIterator.next)
+            rows = SheetData(rows.rowIterator.drop(options.ignoreAfterHeader), rows.resourcesToClose)
```

Comment:
`SheetData` could get a `modifyIterator` method which would return a new `SheetData` with copied iterator and the same `resourcesToClose`.
Then one could do sth. like this:
```scala
rows = rows.modifyIterator(_.drop(options.ignoreAfterHeader)
```